packages/back-end/src/api/members/updateMemberRole.ts

```diff
@@ -0,0 +1,177 @@
+import { isRoleValid, roleSupportsEnvLimit } from "shared/permissions";
+import { cloneDeep } from "lodash";
+import { orgHasPremiumFeature } from "enterprise";
+import { updateOrganization } from "../../models/OrganizationModel";
+import { auditDetailsUpdate } from "../../services/audit";
+import { UpdateMemberRoleResponse } from "../../../types/openapi";
+import { createApiRequestHandler } from "../../util/handler";
+import { updateMemberRoleValidator } from "../../validators/openapi";
+import {
+  Member,
+  OrganizationInterface,
+  ProjectMemberRole,
+} from "../../../types/organization";
+
+function validateRoleAndEnvs(
+  org: OrganizationInterface,
+  role: string,
+  environments?: string[]
+): { memberIsValid: boolean; reason: string } {
+  try {
+    if (!isRoleValid(role, org)) {
+      throw new Error(`${role}) is not a valid role`);
+    }
+
+    if (role === "noaccess" && !orgHasPremiumFeature(org, "no-access-role")) {
+      throw new Error(
+        "Must have a commercial License Key to gain access to the no-access role."
+      );
+    }
+
+    if (environments?.length) {
+      if (!orgHasPremiumFeature(org, "advanced-permissions")) {
+        throw new Error(
+          "Must have a commercial License Key to restrict permissions by environment."
+        );
+      }
+      environments.forEach((env) => {
+        const environmentIds =
+          org.settings?.environments?.map((e) => e.id) || [];
+        if (!environmentIds.includes(env)) {
+          throw new Error(
+            `${env} is not a valid environment ID for this organization.`
+          );
+        }
+      });
+
+      if (!roleSupportsEnvLimit(role, org)) {
+        throw new Error(
+          `${role} does not support restricting access to certain environments`
+        );
+      }
+    }
+  } catch (e) {
+    return {
+      memberIsValid: false,
+      reason: e.message || "Role information is not valid",
+    };
+  }
+
+  return {
+    memberIsValid: true,
+    reason: "",
+  };
+}
+
+export const updateMemberRole = createApiRequestHandler(
+  updateMemberRoleValidator
+)(
+  async (req): Promise<UpdateMemberRoleResponse> => {
+    if (!req.context.permissions.canManageTeam()) {
+      req.context.permissions.throwPermissionError();
+    }
+
+    const orgUser = req.context.org.members.find(
+      (member) => member.id === req.params.id
+    );
+
+    if (!orgUser) {
+      throw new Error("Could not find user with that ID");
+    }
+
+    if (orgUser.managedByIdp) {
+      throw new Error(
+        "This user is managed via an External Identity Provider (IDP) via SCIM 2.0 - User can only be updated via the IDP"
+      );
+    }
+
+    const { member } = req.body;
+
+    const updatedEnvironments = member.environments || orgUser.environments;
+
+    const updatedMember: Member = {
+      ...orgUser,
+      role: member.role || orgUser.role,
+      environments: updatedEnvironments,
+      limitAccessByEnvironment: !!updatedEnvironments.length,
```

Comment:
I think the payload design is fine and makes sense.  The specific situation I'm thinking of is if the current member in Mongo looks like this:

```ts
{
   role: "engineer",
   environments: ["staging"],
   limitAccessByEnvironments: false
}
```

And they call the API with just:

```ts
{
  role: "experimenter"
}
```

I would expect `limitAccessByEnvironments` to remain `false` (keep what's in Mongo because envs aren't being passed into the API).  But because we're spreading in the existing doc and the payload, it will change it to be true.

Reply:
Ah - I see. I thought we reset the `environments` array to an empty array if the `limitAccessByEnvironments` toggled was toggled to "off". 

I can update the logic to:

```ts
    const updatedMember: Member = {
      ...orgUser,
      role: member.role || orgUser.role,
      environments: member.environments || orgUser.environments,
      // If environments are passed in, recalculate limitAccessByEnvironment, otherwise, use existing value
      limitAccessByEnvironment: member.environments
        ? !!member.environments.length
        : orgUser.limitAccessByEnvironment,
    };
```